components/brave_wallet/browser/internal/orchard_sync_state.cc

```diff
@@ -0,0 +1,159 @@
+
+OrchardSyncState::OrchardSyncState(const base::FilePath& path_to_database) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  storage_ = std::make_unique<OrchardStorage>(path_to_database);
+}
+
+OrchardSyncState::~OrchardSyncState() = default;
+
+orchard::OrchardShardTree& OrchardSyncState::GetOrCreateShardTree(
+    const mojom::AccountIdPtr& account_id) LIFETIME_BOUND {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  if (shard_trees_.find(account_id->unique_key) == shard_trees_.end()) {
+    shard_trees_[account_id->unique_key] =
+        orchard::OrchardShardTree::Create(*storage_, account_id);
+  }
+  auto* manager = shard_trees_[account_id->unique_key].get();
+  CHECK(manager);
+  return *manager;
+}
+
+base::expected<OrchardStorage::AccountMeta, OrchardStorage::Error>
+OrchardSyncState::RegisterAccount(const mojom::AccountIdPtr& account_id,
+                                  uint64_t account_birthday_block) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  return storage_->RegisterAccount(account_id, account_birthday_block);
+}
+
+base::expected<OrchardStorage::AccountMeta, OrchardStorage::Error>
+OrchardSyncState::GetAccountMeta(const mojom::AccountIdPtr& account_id) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  return storage_->GetAccountMeta(account_id);
+}
+
+std::optional<OrchardStorage::Error> OrchardSyncState::HandleChainReorg(
+    const mojom::AccountIdPtr& account_id,
+    uint32_t reorg_block_id,
+    const std::string& reorg_block_hash) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  return storage_->HandleChainReorg(account_id, reorg_block_id,
+                                    reorg_block_hash);
+}
+
+base::expected<std::vector<OrchardNote>, OrchardStorage::Error>
+OrchardSyncState::GetSpendableNotes(const mojom::AccountIdPtr& account_id) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  return storage_->GetSpendableNotes(account_id);
+}
+
+base::expected<std::vector<OrchardNoteSpend>, OrchardStorage::Error>
+OrchardSyncState::GetNullifiers(const mojom::AccountIdPtr& account_id) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  return storage_->GetNullifiers(account_id);
+}
+
+std::optional<OrchardStorage::Error> OrchardSyncState::ApplyScanResults(
+    const mojom::AccountIdPtr& account_id,
+    OrchardBlockScanner::Result block_scanner_results,
+    const uint32_t latest_scanned_block,
+    const std::string& latest_scanned_block_hash) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  auto existing_notes = storage_->GetSpendableNotes(account_id);
+  if (!existing_notes.has_value()) {
+    return existing_notes.error();
+  }
+
+  std::vector<OrchardNote> notes_to_add =
+      block_scanner_results.discovered_notes;
+  base::Extend(existing_notes.value(), notes_to_add);
+
+  std::vector<OrchardNoteSpend> nf_to_add;
+
+  for (const auto& nf : block_scanner_results.found_spends) {
+    if (std::find_if(existing_notes.value().begin(),
```

Comment:
`base::ranges::find_if`

Reply:
Fixed

Reply:
Are you sure? I'm still seeing `std::find_if` with a pair of iterators?